### PR TITLE
feat: did deposit

### PR DESCRIPTION
**Pull request by weichweich**
## fixes KILTprotocol/srlabs-audit/issues/7

* reserve deposit when creating a DID
* payback deposit when deleting the DID
* reclaim deposit of a DID
* pay small fee because did blacklisting storage space

## Checklist:

- [ ] I have verified that the code works
  - [ ] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] This PR does not introduce new custom types
  - [ ] If not, I have opened a companion PR with the type changes in the [KILT types-definitions repository](https://github.com/KILTprotocol/type-definitions/pulls)
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: pallets/did/src/did_details.rs [hunk 1]
```diff
@@ -256,14 +257,21 @@ pub struct DidDetails<T: Config> {
...
 impl<T: Config> DidDetails<T> {
 	/// Creates a new instance of DID details with the minimum information,
 	/// i.e., an authentication key and the block creation time.
 	///
 	/// The tx counter is automatically set to 0.
-	pub fn new(authentication_key: DidVerificationKey, block_number: BlockNumberOf<T>) -> Result<Self, StorageError> {
+	pub fn new(
+		authentication_key: DidVerificationKey,
+		block_number: BlockNumberOf<T>,
+		deposit: Deposit<T::AccountId, BalanceOf<T>>,
```
Comment (Diiaablo95) [inline pallets/did/src/did_details.rs:273]:
```suggestion
		deposit: Deposit<AccountIdentifierOf<T>, BalanceOf<T>>,
```

File: pallets/did/src/lib.rs [hunk 2]
```diff
@@ -340,6 +364,10 @@ pub mod pallet {
 		DidAlreadyDeleted,
 		/// An error that is not supposed to take place, yet it happened.
 		InternalError,
+		/// Only the owner of the deposit can reclaim its reserved balance.
+		NotOwnerOfDeposit,
+		/// Only the owner of the deposit can reclaim its reserved balance.
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:369]: 🧐

File: pallets/did/src/lib.rs [hunk 3]
```diff
@@ -700,21 +759,42 @@ pub mod pallet {
...
-				Did::<T>::take(&did_subject).is_some(),
-				Error::<T>::DidNotPresent
-			);
+			Pallet::<T>::delete_did(did_subject)
+		}
 
-			// Mark as deleted to prevent potential replay-attacks of re-adding a previously
-			// deleted DID.
-			DidBlacklist::<T>::insert(&did_subject, ());
+		/// Delete a DID from the chain and all information associated with it,
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:765]: M: Comment inconsistent.

File: pallets/did/src/lib.rs [hunk 4]
```diff
@@ -848,9 +928,9 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
-	// Check if the provided block number is valid,
-	// i.e., if the current blockchain block is in the inclusive range
-	// [operation_block_number, operation_block_number + MaxBlocksTxValidity].
+	/// Check if the provided block number is valid,
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:931]: This is not a public function, why are we making RustDocs for it?

File: pallets/did/src/lib.rs [hunk 5]
```diff
@@ -863,12 +943,12 @@ impl<T: Config> Pallet<T> {
...
-	// Verify the validity of a DID-authorized operation nonce.
-	// To be valid, the nonce must be equal to the one currently stored + 1.
-	// This is to avoid quickly "consuming" all the possible values for the counter,
-	// as that would result in the DID being unusable, since we do not have yet any
-	// mechanism in place to wrap the counter value around when the limit is
-	// reached.
+	/// Verify the validity of a DID-authorized operation nonce.
+	/// To be valid, the nonce must be equal to the one currently stored + 1.
+	/// This is to avoid quickly "consuming" all the possible values for the
+	/// counter, as that would result in the DID being unusable, since we do not
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:949]: Same here

File: pallets/did/src/lib.rs [hunk 6]
```diff
@@ -900,4 +981,23 @@ impl<T: Config> Pallet<T> {
...
 	}
+
+	/// Deletes DID details from storage, adds the identifier to the blacklisted
+	/// DIDs and frees the deposit.
+	fn delete_did(did_subject: DidIdentifierOf<T>) -> DispatchResult {
+		// `take` calls `kill` internally
+		let did_entry = Did::<T>::take(&did_subject).ok_or(Error::<T>::DidNotPresent)?;
+		// *** No Fail beyond this point ***
+
+		kilt_support::free_deposit::<AccountIdentifierOf<T>, CurrencyOf<T>>(&did_entry.deposit);
```
Comment (Diiaablo95) [inline pallets/did/src/lib.rs:992]: Did we say we are not releasing the whole deposit OR making this extrinsic more expensive?

File: pallets/did/src/lib.rs (hunk 4 above)
Comment (weichweich) [reply pallets/did/src/lib.rs:931]: Why do we put a comment in front if it's private? If it's a doc comment you see the comment at the call site when you hover over the function. Private or not this is a useful feature.

File: pallets/did/src/lib.rs (hunk 5 above)
Comment (weichweich) [reply pallets/did/src/lib.rs:949]: same here!

File: pallets/did/src/lib.rs (hunk 6 above)
Comment (weichweich) [reply pallets/did/src/lib.rs:992]: The deposit is a reserved balance. It never leaves the account. So you probably mean that we rather withdraw the fee when we free the deposit instead of withdrawing it immediately?

File: pallets/did/src/lib.rs (hunk 4 above)
Comment (wischli) [reply pallets/did/src/lib.rs:931]: > Why do we put a comment in front if it's private? If it's a doc comment you see the comment at the call site when you hover over the function. Private or not this is a useful feature.

Exactly this!

File: pallets/did/src/lib.rs (hunk 3 above)
Comment (weichweich) [reply pallets/did/src/lib.rs]: ups 1b8df356c07099e1af77fa90ddaacbcb15a3cb0e

File: pallets/did/src/did_details.rs (hunk 1 above)
Comment (weichweich) [reply pallets/did/src/did_details.rs]: 68a40cb3281cb28886f6ffe87bacec65ccd38f19

File: pallets/did/src/benchmarking.rs [hunk 7]
```diff
@@ -102,6 +102,9 @@ benchmarks! {
 
 		let submitter: AccountIdentifierOf<T> = account(DEFAULT_ACCOUNT_ID, 0, DEFAULT_ACCOUNT_SEED);
 
+		let balance = <CurrencyOf<T> as Currency<AccountIdentifierOf<T>>>::minimum_balance() + <T as Config>::Deposit::get() + <T as Config>::Fee::get();
+		<CurrencyOf<T> as Currency<AccountIdentifierOf<T>>>::make_free_balance_be(&submitter, balance);
```
Comment (wischli) [inline pallets/did/src/benchmarking.rs]: C: Move this to a function for easier readability, e.g., `fn airdrop(who: T::AccountId)`.

File: pallets/did/src/lib.rs [hunk 8]
```diff
@@ -423,6 +450,16 @@ pub mod pallet {
 
 			let did_identifier = details.did.clone();
 
+			// TODO: how to ensure reserve AND withdraw!?
```
Comment (wischli) [inline pallets/did/src/lib.rs:453]: S: Remove the TODO. 

Your solution of checking the reservability of `Deposit + Fee` seems fine to me.
```suggestion
```

File: pallets/did/src/lib.rs [hunk 9]
```diff
@@ -437,10 +474,31 @@ pub mod pallet {
...
+					owner: sender.clone(),
+					amount: T::Deposit::get(),
+				},
+			)
+			.map_err(Error::<T>::from)?;
+
+			// *** No Fail beyond this call ***
+			CurrencyOf::<T>::reserve(&did_entry.deposit.owner, did_entry.deposit.amount)?;
+
+			// withdraw the fee, we made sure that enough balance is available. But if this
```
Comment (wischli) [inline pallets/did/src/lib.rs:490]: S: Start with uppercase

```suggestion
			// Withdraw the fee. We made sure that enough balance is available earlier. But if this
```

File: support/src/lib.rs [hunk 10]
```diff
@@ -17,5 +17,17 @@
...
 #![cfg_attr(not(feature = "std"), no_std)]
 
+use deposit::Deposit;
+use frame_support::traits::{Currency, ReservableCurrency};
+use sp_runtime::traits::Zero;
+
 pub mod deposit;
 pub mod traits;
+
+pub fn free_deposit<A, C>(deposit: &Deposit<A, C::Balance>)
```
Comment (wischli) [inline support/src/lib.rs:27]: C: Use this function in other pallets as well. But I suppose this will be handled in a separate PR.

**Review by wischli — APPROVED** [review]
I only have a few minor notes but overall looks good to me!

File: pallets/did/src/lib.rs (hunk 4 above)
Comment (Diiaablo95) [reply pallets/did/src/lib.rs:927]: Aren't they then also exported as RustDoc?

Comment (weichweich) [reply pallets/did/src/lib.rs:926]: Rust is smart. Rust doesn't show private functions in it's doc. Be like Rust!
